File: Data Structures & Algorithms/invert-a-binary-tree/submission-5.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
   public:
    TreeNode* invertTree(TreeNode* root) {
        // Return nullptr if the root is nullptr
        if (root == nullptr) {
            return nullptr;
        }

        queue<TreeNode*> q;
        q.push(root);

        // Exit the while loop when the queue is empty
        while (!q.empty()) {
            // Get the node from the queue to process
            // and remove that node form the queue
            TreeNode* node = q.front();
            q.pop();

            // Swap the node's left and right children
            swap(node->left, node->right);

            // Push the swapped left and right children into the queue
            // for next procesing iteration in the while loop
            // Note that we only push existed child into the queue
            // because there's nothing to process if that child node doesn't exist
            if (node->left != nullptr) {
                q.push(node->left);
            }
            if (node->right != nullptr) {
                q.push(node->right);
            }
        }
        // Return the modified tree's root node
        return root;
    }
};

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        curr = root
        s = []

        # Exit the loop when current node is empty and the stack is empty
        while curr or len(s) != 0:
            # Exit the loop when current node is nullptr,
            # otherwise, push the current node into stack
            # and keep going left
            while curr:
                s.append(curr)
                curr = curr.left

            # 1. Get current node from the top of the stack
            # 2. Remove the visited node since we don't need it anymore
            # 3. Decrement k
            curr = s.pop()
            k -= 1

            # If k is 0, that mean we've reached our goal, return the current node's value
            if k == 0:
                return curr.val

            # Move the current node to its right
            curr = curr.right

        # Return -1 to satisfy the compiler
        # Note that we'll never reach this because 1 <= k <= n <= 10^4
        return -1
